sdks/python/apache_beam/internal/pickler.py

```diff
@@ -33,12 +33,16 @@
 import base64
 import logging
 import sys
+import threading
 import traceback
 import types
 import zlib
 
 import dill
 
+# Pickling, especially unpickling, can cause broken module imports
+# if executed concurrently, see: BEAM-8651.
+pickle_lock = threading.Lock()
```

Comment:
Should this be RLock?

Reply:
I considered using RLock here and went for Lock since it is supposed to be faster on Python 2 and I didn't see a scenario where a call to pickler.loads() or pickler.dumps() would initiate another call to these methods in the same thread. But perhaps it is possible to craft a pickled payload, which references methods of pickler.py, and dill would call these methods during unpickling?

Reply:
Turns out that multiple calls to pickle ops are possible within a thread: dill does module imports; when some modules create beam objects in a global scope,  beam objects themselves call pickle ops to verify that they are pickleable. This causes a deadlock that can be avoided with a reentrant lock, sending https://github.com/apache/beam/pull/10265.